README.md
BMS Cell Node Firmware
======================

[![](https://gitlab.com/kroesche/bmsnode/badges/master/pipeline.svg)](https://gitlab.com/kroesche/bmsnode/pipelines)
[![](https://gitlab.com/kroesche/bmsnode/badges/master/coverage.svg)](https://gitlab.com/kroesche/bmsnode/-/jobs/artifacts/master/file/test/reports/bmstest-coverage.html?job=test)
[![](https://gitlab.com/kroesche/bmsnode/-/jobs/artifacts/master/raw/test/reports/check-badge.svg?job=test)](https://gitlab.com/kroesche/bmsnode/-/jobs/artifacts/master/file/test/reports/check/index.html?job=test)

* * * * *

This repository holds firmware for **BMSNode** BMS cell boards. BMSNode boards
are used to monitor a single cell or set of cells connected in parallel, in a
battery pack. It also includes a shunt resistor that can be used to help with
cell balancing.

This device was heavily inspired by the
[diyBMSv4 board](https://github.com/stuartpittaway/diyBMSv4). However this
firmware was written completely from scratch and does not use any of the
original design or protocol.

The BMSNode board design repository
[can be found here](https://gitlab.com/kroesche/bmsnode_board).

License
-------

[![](doc/img/license-badge.svg)](https://opensource.org/licenses/MIT)

This project uses the [MIT license](https://opensource.org/licenses/MIT).
See the [LICENSE.txt](LICENSE.txt) file for the license terms.

Repo Layout
-----------

```
.
├── build   (scripts for building and other automation)
├── doc     (various documentation and API docs generation)
├── src     (firmware source code)
├── test    (test code and scripts)
└── util    (tools to help with usage and test)
```

Documentation
-------------

The `doc` directory contains most project documentation such as specifications.
Otherwise see the various README files and as a last resort see comments in
script files. Most Makefiles have a `make help` target.

The firmware source code is documented using doxygen-style comments. Generated
documentation [can be found here](https://kroesche.gitlab.io/bmsnode/html/index.html).

Tools and Environment
---------------------

Assumes a POSIX-ish command line environment and shell like bash. Almost
everything is done from the command line, and using Makefiles. The firmware is
compiled with `avr-gcc`.



CHANGELOG.md
# CHANGELOG

**BMSNode Firmware**

<!--- next entry here -->

## 0.11.0
2021-05-24

### Breaking changes

#### Switch to attiny1614 MCU ... (2a55aa46d9d8e92b8f2dc927e6d3d6e5ab17419f)

Port to attiny1614 for new v3 boards.

#### large update throughout to support v3 boards with ATtiny1614 (e30d046c8e1ebe2f0c9a8ca7753367f0e07b334d)

port to ATtiny1614 for new v3 boards

- update firmware for new 1614 peripherals
- update unit test code to match
- update host utilities and provisioning
- update docker image for new tools
- runs on v3 board

### Features

- add ext and mcu temp readings #59 #63 (13a85c2701a2d53838f47fefd0708d9c6a62a956)
- **test:** update unit tests for new firmware (234fbc9a661e0e5253d416ebc3021d7cd6a5060e)

### Fixes

- bmsloader dont fail verify due to boot vector [skip ci] (54ebcb713d245d6ee85e9a71c3c48d365cf2591d)
- update doxygen theme, no functional changes (59a458143eadd1e3a4ab2efe0a7428d2f54642e6)
- parser 5 sec timeout if partial packet rcvd #53 (1ef0f593e5e0951020332dc2a4b3f6ad89fb35cb)
- add watchdog for all states #57 (ec0142371ed20600758fafd520ef471fcb175d23)
- add code size check and todo report #16 #23 (03cab9428c2535adda2ccccb1ad3ce2c8b56be7c)
- **bootloader:** add optiboot_x bootloader for 1614 mcu [skip ci] (93403db634dcb227a4340eefd0ffbf9192e46182)
- **bootloader:** clean up bootloader files location [skip ci] (3dea763c7cf635d99827d9fe64ddd64f67260f21)
- **bootloader:** reorg bl files, add optiboot_x prep for upcoming 1614 (d636d9f268b892582a42edac20d26f987fbf282e)
- **build:** update docker to support attiny1614 build tools (f8632120b6fafb8a97b092e977a2c38ddf63a43f)
- update provisioning for 1614, update bmsutil for optiboot_x (0c4dd1c2d61cd927362e1cc5e51e32b5cdf68130)
- **adc:** update ADC for v3 board/new MCU (175ae308bd40bbbf4978699331f658d559c7543c)
- adjust pwm freq and set gpio for lower power (15b64e6fb7bd172e4cc5029d2963972530b66641)
- update docs for new v3 design (a6a2718c1aaa3abd8cf070987b76fcebbab45fe1)

## 0.10.3
2020-12-13

### Fixes

- clarify SETPARM/ADDR behavior #49 (1ff1358f7e3bff6137f6228bb37dbe225f90a680)
- move ADC timing to ADC module #60 (8825875b91f37304cfcf5e08ffd34e2c457a6136)
- **cfg:** add factory reset command #39 (9f9788570d55152e838db2c6d798e3dbfc92f58c)
- **shunt:** shunt PWM adjustment smoothing #56 (9bd68c4ddca2049f27124d9d1d428418fa753e0e)
- **adc:** add ADC smoothing #55 (7e6515c5aa090293eaef1981e48ef302ce49cd6b)

## 0.10.2
2020-11-25

### Fixes

- error in ADC sample timing (7d2f162bb28aeae67cbf1d377a4c83568e7696ef)

## 0.10.1
2020-11-24

### Fixes

- reduce PWM freq to about 500 Hz (12f52fc66e7e79d5671e6a52fa1c08ec3890619c)

## 0.10.0
2020-11-20

### Features

- add LED abstraction (83b7cf23c54677f906d4fe7183a3ea36e00363f9)
- add list helper module (192158985f4fd856bf4a817ad04fa97d9401f342)
- add scheduled timers (b97bfb179dad8c550958e9efee8824fbba3d0272)
- change rate to 9600 [skip ci] (4c648204e52240b196c83ebb81f44e276505d6af)
- **shunt:** use PWM for shunt with temp limit #46 ... (6b1fc0c2f4c6c314012abe22d0b141fecc64ca70)

### Fixes

- add new v2.0 boards to utils (5ea10c7d8377e02a46e3715a97ae2d6ccb02eb1d)
- update bootloader hex files with latest optiboot version [skip ci] (786e85ef6c7f703e1141ca28171034af07e43c82)
- parameters not being stored #54 (707c34b3d10cb25dc1c8858e91fb9a1d7935cb54)
- reduce app startup blink timeout from 5 to 1 seconds (850e9c3c3ec15ff5742023f1b71a9fe7285a3fa9)
- convert main loop to state machine (00803927f8f014e3f7a9dac955dcea90e9b61e93)
- clean up and improve documentation, add gitlab pages (1f392fb8b6c6dc6547fc4d928a29b888be3119ee)
- clean up some cppcheck warnings (36d594b48561e57a2503d5ad2428e3d98ad5218b)

## 0.9.0
2020-09-17

### Features

- change serial rate to 2400, update tools and docs (05e7fe85ff92970f6a9ae47a97dd1c3b53f979d3)

## 0.8.1
2020-09-16

### Fixes

- use cal constants for cell V calc, update cal value #32 #44 #50 (92601191dac303aa6651154281c67befc17c60d6)
- make bmsutil a submodule [skip ci] (0bc446185f1eef7274c4c8cea69182d6ed73a12e)
- add dumpflash target to Makefile for debug [skip ci] (61b0ae45a693d68ccbb2aaf0c9e8823c94fbf7aa)
- add testmode command to firmware ... (e7625ff7bc19fb77d43c023883b7f174b95fab7b)
- update thermistor table to match current part #38 #43 ... (f7b851eb8cfab0656632c9a65a03f7318fbf9c14)
- add visual ident to PING command #7 (59929b1f7b6f96cc5cc3b2769c13c88f6bbc87cf)

## 0.8.0
2020-09-02

### Features

- **board:** add support for new board bmsnode v1.0.0 (e7bcb5206ca2604bc4f03009bcdec424aaa0cb00)
- **board:** add support for JLC v1.0.0 boards #47 (cfa3f25e4701834b903ab26a07c0eb28c9a965db)

### Fixes

- minor doc fixes for board update [skip ci] (eff6563e879645cfdb4744d773ab87194aa51ef5)
- update bootloader README for half-duplex diffs [skip ci] (dea194e26b3e052c9d260fa49ee774623aa7b07d)

## 0.7.0
2020-05-30

### Features

- **cfg:** add config set/get command, new set of parameters ... (f02b16586df129e5a2be71e425a7bea430e37e8d)
- **shunt:** add shunt config, automation, status reporting #10 (b663812c11168afc410fb09ca5a767e5e84b7e4f)

## 0.6.1
2020-05-20

### Fixes

- add doxygen docs generation, no functional change (#35) (19967ce43d8aff563b3bc85697bfbb9f2820b777)

## 0.6.0
2020-05-15

### Features

- **shunt:** add shunt mode and command #11 (a8a202c0139a4c87c40041e19544bcace7367f80)

### Fixes

- **cmd:** add api to get last command processed ... (a88d2a6f46f03f2021295008bcc9e85d8d4e0b49)

## 0.5.0
2020-05-12

### Features

- **adc:** add ADC module and raw data command #30 (f54f3d8220b40a6d46f51ec2dd0f208da26961f3)
- **adc:** add cell voltage reading and status command #8 (eb28af8a6532122f4918185804e6b2a0c583f189)
- **adc:** add temp sensor reading to status command #9 (99f7ec45b99799685282a7c15080c55ec3ed6728)

## 0.4.0
2020-05-10

### Features

- **power:** add power save mode in main loop #13 (968f9676155ddf031cdf633a50de8d30c2b0b5c0)

### Fixes


src/adc.c
/******************************************************************************
 * SPDX-License-Identifier: MIT
 *
 * Copyright 2020 Joseph Kroesche
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 *****************************************************************************/

#include <stdint.h>
#include <stdbool.h>

#include <avr/io.h>

#include "iomap.h"
#include "thermistor_table.h"
#include "cfg.h"
#include "tmr.h"
#include "adc.h"

// how often to take a sample set
#define ADC_SAMPLE_PERIOD 100

// smoothing filter weight, out of 32
// 8 ==> smoothing constant of 0.25
#define FILTER_WEIGHT 8

// convenience macro to check if ADC is enabled
// only check ADC1 but assume ADC0 and 1 track
#define ADC_ENABLED (ADC1.CTRLA & 1)

// channel map - mux channels to sample
#define NUM_CHANNELS 4
static const struct
{
    ADC_t *padc;
    uint8_t muxpos;
    uint8_t refsel;
} channels[NUM_CHANNELS] =
        // note for vsense mux is 7 for adc0 and 3 for adc1
    {   { &ADC1, 3, ADC_REFSEL_INTREF_gc },     // VSENSE
        { &ADC0, 4, ADC_REFSEL_VDDREF_gc },     // TSENSE
        { &ADC0, 11, ADC_REFSEL_VDDREF_gc },    // EXTTEMP
        { &ADC0, 0x1E, ADC_REFSEL_INTREF_gc }   // MCU temp sensor
    };

// storage for sample data
static uint16_t results[4];

// adc sample timer
static uint16_t adc_timeout;

//////////
//
// See header file for public function API descriptions.
//
//////////

// ADC timing notes
//
// ADC clock should be 50-1500 kHz, per datasheet
// Given input clock is 10 MHz, choose /16 prescaler
// 10/16 ==> 625 kHz --> 1.6 uS cycle time
//
// The normal conversion is 2 clocks sample time plus 13 clocks conversion,
// for a total of 15 ADC cycles or 24 uS. To try and improve accuracy, the
// sampel time is further extended by another 8 cycles for a total sample
// time of 10 cycle or 16 uS. Now, total conversion time is 10+13=23 cycles
// or 36.8 uS.
//
// initialize and power up ADC circuits, and set up references. This must be
// called before using adc_sample().
// Because different references are needed, both ADC0 and ADC1 are used, each
// using a different reference.
void adc_powerup(void)
{
    // disable digitial inputs for the analog inputs
    // TODO: should this be done at system init?
    PORTA.PIN4CTRL = PORT_ISC_INPUT_DISABLE_gc;
    PORTA.PIN7CTRL = PORT_ISC_INPUT_DISABLE_gc;
    PORTB.PIN0CTRL = PORT_ISC_INPUT_DISABLE_gc;

    // turn on the GPIO used as external supply for resistor dividers
    REFON_PORT.OUTSET = REFON_PIN;

    // Set up VREF to provide 1.1V ref for ADC0 and 2.5V for ADC1
    // ADC1 will be used for measuring the voltage and needs 2.5 reference
    // ADC0 will be used for measuring the two temp sensors and the internal
    // temp sensor. The external temp sensors need VDD reference and the
    // MCU temp sensor need 1.1V reference.
    VREF.CTRLA = VREF_ADC0REFSEL_1V1_gc;
    VREF.CTRLC = VREF_ADC1REFSEL_2V5_gc;

    // set up ADC1 which is used for voltage measurement
    ADC1.CTRLC = ADC_SAMPCAP_bm         // reduced sampcap
               | ADC_REFSEL_INTREF_gc   // internal reference
               | ADC_PRESC_DIV16_gc;    // prescaler /16
    ADC1.SAMPCTRL = 8;                  // add 8 more sampling cycles
    ADC1.CTRLA = 1;                     // enable ADC

    // set up ADC 0 which is used for temperature measurements
    // we set reference here, but it needs to change dynamically depending
    // on which temp is being measured
    ADC0.CTRLC = ADC_SAMPCAP_bm         // reduced sampcap
               | ADC_REFSEL_VDDREF_gc   // VDD reference
               | ADC_PRESC_DIV16_gc;    // prescaler /16
    ADC0.SAMPCTRL = 8;                  // add 8 more sampling cycles
    ADC0.CTRLA = 1;                     // enable ADC

    // reset the ADC timeout
    adc_timeout = tmr_set(3);
}

// shut down ADC and external ref
void adc_powerdown(void)
{
    // shut down the ADC and turn off the external reference
    ADC0.CTRLA = 0;
    ADC1.CTRLA = 0;
    REFON_PORT.OUTCLR = REFON_PIN;
}

// exponential smoothing of the ADC reading
static uint16_t adc_filter(uint16_t sample, uint16_t smoothed)
{
    smoothed = (sample * FILTER_WEIGHT) + (smoothed * (32 - FILTER_WEIGHT));
    smoothed = (smoothed + 16) / 32;
    return smoothed;
}

// collect one sample from the specific ADC using parameters
// does not filter
// Per timing notes above, a conversion takes about 40 uS and we do two here.
// So this function takes about 80 uS and is blocking.
uint16_t adc_sample(ADC_t *padc, uint8_t muxpos, uint8_t refsel)
{
    padc->MUXPOS = muxpos;          // select channel
    padc->CTRLC = ADC_SAMPCAP_bm | ADC_PRESC_DIV16_gc
                | refsel;           // select reference
    padc->COMMAND = 1;              // start conversion
    while (!(padc->INTFLAGS & ADC_RESRDY_bm))
    {}      // wait for conversion complete

    // throw away the first result and run the conversion again.
    // this enforces a settling time after the mux change
    // NOT SURE IF THIS IS REALLY NECESSARY - NEED TO DO TESTING
    // TODO: look into '1614 sample accumulation feature
    // and 1614 sample delay feature
    padc->INTFLAGS = ADC_RESRDY_bm; // clear ready flag
    padc->COMMAND = 1;              // start conversion
    while (!(padc->INTFLAGS & ADC_RESRDY_bm))
    {}      // wait for conversion complete

    // return the reading, filtering to be done by caller
    return padc->RES;
}

// make one pass through all the samples, filter, and save.
// There are 4 channels and each call to adc_sample() is about 80 uS, so
// this function will take about 320 uS and is blocking.
void adc_collect(void)
{
    // loop to read all the channels and store results
    for (uint8_t idx = 0; idx < NUM_CHANNELS; ++idx)
    {
        uint16_t result = adc_sample(channels[idx].padc,
                                     channels[idx].muxpos,
                                     channels[idx].refsel);
        results[idx] = adc_filter(result, results[idx]);
    }
}

// run adc sampling at periodic interval
void adc_run(void)
{
    if (ADC_ENABLED && tmr_expired(adc_timeout))
    {
        adc_timeout += ADC_SAMPLE_PERIOD;
        adc_collect();
    }
}

// return the raw data in an array
uint16_t *adc_get_raw(void)
{
    return results;
}

// return the cell voltage in millivolts
uint16_t adc_get_cellmv(void)
{
    uint32_t mv1024 = (uint32_t)results[0] * (uint32_t)g_cfg_parms.vscale;
    mv1024 /= 1024U;
    mv1024 += g_cfg_parms.voffset;
    return (uint16_t)mv1024;
}

// return the thermistor temperature in C
int16_t adc_get_tempC(enum adc_channel ch)
{
    if (ch == ADC_CH_MCU_TEMP)
    {
        // algorithm from 1614 data sheet
        int8_t offset = SIGROW.TEMPSENSE1;
        uint8_t gain = SIGROW.TEMPSENSE0;
        uint32_t mcutemp = results[ch] - offset;
        mcutemp *= gain;
        mcutemp += 0x80;    // half bit rounding
        mcutemp >>= 8;      // Kelvin
        mcutemp -= 273;
        return (int16_t)mcutemp;
    }
    else
    {
        return adc_to_temp(results[ch]);
    }
}
